### DISCUSSION TITLE
Proper way to get more data in catchers

### DISCUSSION THREAD

**Discussion by TPausL** [rwf2/Rocket#2508] [Questions]
**Existing Functionality**
I want to pass details to catchers. From a Guard there is the request.local_cache, which works but is sometimes error prone to set up as it has to typed correctly on both ends (setting and retrieving) and has to wrapped by an option. But the main problem with that is that there is no way of setting it within a request handler itself. I for example validate my incoming data in a request guard but then store it in the database later in the request handler. If that fails i can only manullay return an error instead of say passing a string and then sending a generic response with that custom message, which would reduce code duplication.

**Suggested Changes**
some way to store data in the request more conveniently, which is accesible from every part of the code touching the request somehow.

**Alternatives Considered**
being able to write to the request local_cache within request handler

**Additional Context**
if you want to have a look at my code in order to understand the problem better or suggest better ways feel free to take a look:
[TPausL/power-music-backend](https://github.com/TPausL/power-music-backend)

**Comment by SergioBenitez** [comment]
What we eventually want is #749, and it's something I intend to work in soon™. I _think_ that would resolve your issue? I'm not sure I totally understand the motivation behind it.

> I for example validate my incoming data in a request guard but then store it in the database later in the request handler. If that fails i can only manullay return an error instead of say passing a string and then sending a generic response with that custom message, which would reduce code duplication.

Why can't you place the error in request-local cache in the request guard implementation?

I'm converting this into a discussion, assuming #749 would indeed be the desired fix.

**Reply by Atlinx** [reply]
Are we able to put errors in the request-local cache? I thought only strings were able to be cached since the value has to implement the sealed `Shareable` trait. I'm also facing this issue right now with error handling in a guard and I can't find a good way to solve it.

**Reply by Atlinx** [reply]
Ah, turns out I had to wrap my cached object inside of a `Vec`

Here's a snippet of how I'm using it. I'm using my own Responder called `APIResponse` which lets you return JSON data and a status code. This doubles as an error that guards can return when failing):
```rust
// Inside fn from_data for a `FromData` implementation
        ...
        let user_login =
            try_outcome!(Json::<UserLogin>::from_data(req, data)
                .await
                .map_failure(|_| {
                    APIResponse::new_message(
                        Status::UnprocessableEntity,
                        "Error while parsing user login.",
                    )
                    .cache(req)
                })) as Json<UserLogin>;
        ...

// Inside APIResponse
    ...
    pub fn cache(self, req: &Request) -> APIResponse {
        req.local_cache(|| vec![self.clone()]);
        self
    }
    ...

// Catcher that takes any cached errors and returns it
#[catch(default)]
fn default_handler(status: Status, req: &Request) -> APIResponse {
    let response: &Vec<APIResponse> = req.local_cache(|| vec![bad_request()]);
    response[0].clone()
}
```
